### PR TITLE
feat: add a function to get number of logical cpu cores

### LINKED ISSUE #4879: feat: add a function to get number of logical cpu cores
## Feature Request

**Is your feature request related to a problem? Please describe:**

Now we can only get number of physical cpu cores by `dbms/src/Common/getNumberOfPhysicalCPUCores.cpp`.
1. `getNumberOfPhysicalCPUCores()` will return number of logical cpu cores in some cases which is misleading.
2. In most cases, number of logical cpu cores matters rather than physical cpu cores especially in cloud env.

**Describe the feature you'd like:**

Add a function to get number of logical cpu cores which support different env like docker/k8s, cgroup subsystem.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: close #4879 

Problem Summary:

### What is changed and how it works?
Add a function to get number of logical cpu cores:
1. Normally, `$cpu = std::thread::hardware_concurrency()`
2. With cgroup, `$cpu = min(len(cpuset), cgroup_quota / cgroup_period)`

detail doc: https://github.com/pingcap/tiflash/issues/4879#issuecomment-1131378727
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Lloyd-Pottiger** [comment]
- [ ] add unit tests (if gtest support)
- [ ] test on k8s/docker env

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+namespace DB::ErrorCodes
+{
+extern const int CPUID_ERROR;
+} // namespace DB::ErrorCodes
+
+#if defined(__linux__)
+// Try to look at cgroups limit if it is available.
+static inline int read_int_from(const char * filename, int default_value)
+{
+    std::ifstream infile(filename);
```
Comment (ywqzzy) [inline dbms/src/Common/getNumberOfLogicalCPUCores.cpp:34]: nit: use poco lib to read file?

Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfLogicalCPUCores.cpp:34]: `Poco::FileInputStream` will throw an exception when file no found directly.

**Comment by Lloyd-Pottiger** [comment]
Test on docker with `--cpus=$n`, the result of `getNumberOfLogicalCPUCores` equal to `$n`.

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    if (enabled_v2)
+    {
+        std::string cpu_filter = "0::";
+        std::ifstream cgroup_cpu_info("/proc/self/cgroup");
+        if (cgroup_cpu_info.is_open())
+        {
+            std::string line;
+            while (std::getline(cgroup_cpu_info, line))
+            {
+                std::string::size_type cpu_str_idx = line.find(cpu_filter);
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfLogicalCPUCores.cpp]: the return type of `std::string::find` is `std::string::size_type`

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+unsigned getCGroupLimitedCPUCores(unsigned default_cpu_count)
+{
+    std::string cgroup_controllers = "/sys/fs/cgroup/cgroup.controllers";
+    std::ifstream cgroup_controllers_info(cgroup_controllers);
+    // If cgroup.controllers is open, we assume we are running on a system with cgroups v2
+    // Otherwise v1
+    bool enabled_v2 = cgroup_controllers_info.is_open();
+
+    if (enabled_v2)
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfLogicalCPUCores.cpp]: done~

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        }
+    }
+    return getCGroupDefaultLimitedCPUCores(default_cpu_count);
+}
+#endif // __linux__
+
+unsigned getNumberOfLogicalCPUCores()
+{
+    unsigned logical_cpu_count = std::thread::hardware_concurrency();
+    if (logical_cpu_count == 0)
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfLogicalCPUCores.cpp]: good idea! thank u!

File: dbms/src/Common/getNumberOfLogicalCPUCores.h [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+/// Get number of CPU cores without hyper-threading.
+unsigned getNumberOfLogicalCPUCores();
```
Comment (JaySon-Huang) [inline dbms/src/Common/getNumberOfLogicalCPUCores.h:18]: show it in bootstrap https://github.com/pingcap/tiflash/blob/e03dcc1e38e14906e7598c7e8775a2660e9ab58a/dbms/src/Server/Server.cpp#L1403-L1411

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+#define PER_CPU_SHARES 1024
+    unsigned share_count = default_cpu_count;
+    if (cgroup_share > -1)
+    {
+        share_count = ceil(static_cast<float>(cgroup_share) / static_cast<float>(PER_CPU_SHARES));
+    }
+
+    return std::min(default_cpu_count, std::min(share_count, quota_count));
+}
```
Comment (JaySon-Huang) [inline dbms/src/Common/getNumberOfLogicalCPUCores.cpp:136]: https://github.com/ClickHouse/ClickHouse/pull/35815

Checkout whether we should adapt to these changes pls

File: dbms/src/Common/getNumberOfLogicalCPUCores.h (hunk 5 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfLogicalCPUCores.h:18]: I will do that in another pr to fix #3821

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp (hunk 6 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfLogicalCPUCores.cpp:136]: I think it is reasonable, `cpu.shares` only works when the machine is free, so cpus should be visible to tiflash. I will adapt it soon.

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    // The maximum bandwidth limit. It's in the following format:
+
+    // $MAX $PERIOD
+
+    // which indicates that the group may consume upto $MAX in each $PERIOD duration.
+    // "max" for $MAX indicates no limit. If only one number is written, $MAX is updated.
+    std::ifstream infile(filename);
+    if (!infile.is_open())
+    {
+        return std::make_pair(1, 1);
```
Comment (ywqzzy) [inline dbms/src/Common/getNumberOfLogicalCPUCores.cpp:102]:
```suggestion
        return {1, 1};
```

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    }
+    std::string quota;
+    int period;
+    infile >> quota >> period;
+    if (quota == "max")
+    {
+        return std::make_pair(period, period);
```
Comment (ywqzzy) [inline dbms/src/Common/getNumberOfLogicalCPUCores.cpp:109]: ditto

Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfLogicalCPUCores.cpp]: done

**Comment by Lloyd-Pottiger** [comment]
It does not work in some case…… I will write a doc and do more tests.

**Comment by breezewish** [comment]
Is it easy to add tests for it? How about setting up a docker, or more directly, just use cgroup in tests?

**Comment by Lloyd-Pottiger** [comment]
> Is it easy to add tests for it? How about setting up a docker, or more directly, just use cgroup in tests?

Currently, I did not find the way to test by gtests. I tested manually before. (I did not test on cgroup v2 because I still could not set up the environment after trying a lot on my own machine) Do you mean run a docker in a unit test? Maybe it is hard, and is not a good way which means we may need to introduce more dependencies. However, there are some bugs indeed. I will fix later.

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    std::ifstream infile(filename);
+    if (!infile.is_open())
+    {
+        return default_value;
+    }
+    int idata;
+    if (infile >> idata)
+        return idata;
+    else
+        return default_value;
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfLogicalCPUCores.cpp]: fixed!

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+    if (cgroup_quota > -1 && cgroup_period > 0)
+    {
+        quota_count = ceil(static_cast<float>(cgroup_quota) / static_cast<float>(cgroup_period));
+    }
+
+    return std::min(cpuset_count, quota_count);
+}
+
+static unsigned read_cpuset_count_from(const char * filename, unsigned default_value)
```
Comment (jiaqizho) [inline dbms/src/Common/getNumberOfLogicalCPUCores.cpp:67]: Would you like to add some `gtests` to test all of functions？

You can find some example in `man`.

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+    size_t first = 0;
+    while (first < line.size())
+    {
+        size_t last = line.find(',', first);
+        if (last == std::string::npos)
+        {
+            last = line.size();
+        }
+        std::string cpu_set = line.substr(first, last - first);
+        size_t dash = cpu_set.find('-');
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfLogicalCPUCores.cpp:96]: seems it is not allowed to write notes
```
$ echo "0-3 # 0-3" >cpuset.cpus
bash: echo: write error: Invalid argument
```

File: dbms/src/Common/getNumberOfLogicalCPUCores.cpp (hunk 10 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfLogicalCPUCores.cpp:67]: I will

**Pull request closed without merging**